presentation/frontend/src/presentation-frontend/StateTracker.ts

```diff
@@ -47,77 +47,68 @@ export class StateTracker {
     const hierarchy = this._expandedHierarchies.get(rulesetId);
     if (!hierarchy)
       return;
 
     const removedKeys: NodeKey[] = [];
-    for (const collapsedNode of collapsedNodes) {
-      const existingNode = hierarchy.get(collapsedNode.id);
-      if (!existingNode)
-        continue;
-
-      // remove current source from the list of sources that have this node expanded
-      existingNode.expandedIn.delete(sourceId);
-      // if there are other sources that have this node expanded leave it
-      if (existingNode.expandedIn.size !== 0)
+    for (const [nodeId, expandedNode] of hierarchy) {
+      expandedNode.expandedIn.delete(sourceId);
+      // if there are other sources that have this node expanded leave it.
+      if (expandedNode.expandedIn.size !== 0)
         continue;
 
-      hierarchy.delete(collapsedNode.id);
-      removedKeys.push(collapsedNode.key);
+      hierarchy.delete(nodeId);
+      removedKeys.push(expandedNode.key);
     }
 
-    // if hierarchy does not have any nodes remove it
     if (hierarchy.size === 0)
       this._expandedHierarchies.delete(rulesetId);
 
     await this.updateHierarchyStateIfNeeded(imodel.key, rulesetId, "nodesCollapsed", removedKeys);
   }
 
-  public async onHierarchyClosed(imodel: IModelConnection, rulesetId: string, sourceId: string) {
-    const hierarchy = this._expandedHierarchies.get(rulesetId);
-    if (!hierarchy)
+  public async onExpandedNodesChanged(imodel: IModelConnection, rulesetId: string, sourceId: string, expandedNodes: NodeIdentifier[]) {
+    let hierarchy = this._expandedHierarchies.get(rulesetId);
+    if (expandedNodes.length === 0 && !hierarchy)
       return;
+    if (!hierarchy) {
+      hierarchy = new Map<string, ExpandedNode>();
+      this._expandedHierarchies.set(rulesetId, hierarchy);
+    }
 
     const removedKeys: NodeKey[] = [];
-    for (const [nodeId, expandedNode] of hierarchy) {
-      expandedNode.expandedIn.delete(sourceId);
-      // if there are other sources that have this node expanded leave it.
-      if (expandedNode.expandedIn.size !== 0)
+    const addedKeys: NodeKey[] = [];
+    for (const [key, existingNode] of hierarchy) {
+      // existing node is in new expanded nodes list. Add current source
+      if (expandedNodes.find((expandedNode) => expandedNode.id === key)) {
+        existingNode.expandedIn.add(sourceId);
         continue;
+      }
 
-      hierarchy.delete(nodeId);
-      removedKeys.push(expandedNode.key);
+      // node was not found in expanded nodes list. Remove current source
+      existingNode.expandedIn.delete(sourceId);
+      if (existingNode.expandedIn.size !== 0)
+        continue;
+
+      removedKeys.push(existingNode.key);
+      hierarchy.delete(key);
     }
 
-    if (hierarchy.size === 0)
-      this._expandedHierarchies.delete(rulesetId);
+    // add any new nodes that were not in expanded nodes hierarchy already
+    for (const expandedNode of expandedNodes) {

```

Comment:
Not sure if we really have to worry about a large size of `expandedNodes` list, but... I think we could:
- have a list of handled `expandedNodes` indexes
- fill it while iterating the loop above this
- using that information, immediately skip the node in this loop since we know it was handled

This would allow us to avoid unnecessarily doing `hierarchy.get(expandedNode.id)`.

Reply:
I don't think its worth doing this micro optimization as there should not be very large arrays of `expandedNodes` passed in because I don't think user would expand thousands of nodes in single tree.